### ISSUE TITLE
bug: ingress controller pod crashes when ingress is deleted

### ISSUE THREAD

**Issue by alf5** [apache/apisix-ingress-controller#941]
### Issue description

Pod crashes when ingress resource is deleted (i.e. kubectl delete ingress http-bin).


### Environment

- your apisix-ingress-controller version (output of apisix-ingress-controller version --long):
```
Version: 1.4.0
Git SHA: no-git-module
Go Version: go1.16.12
Building OS/Arch: linux/amd64
Running OS/Arch: linux/amd64
```
- your Kubernetes cluster version (output of kubectl version): `1.21.1` - running in a kind cluster

- if you run apisix-ingress-controller in Bare-metal environment, also show your OS version (uname -a): 
`Linux dev-amd 5.13.0-37-generic #42-Ubuntu SMP Tue Mar 15 14:34:06 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

### Minimal test code / Steps to reproduce

1. Follow https://apisix.apache.org/docs/ingress-controller/practices/proxy-the-httpbin-service-with-ingress
2. execute - kubectl delete ingress httpserver-ingress

### Actual result

apisix ingress controller pod crashes.

Kubernetes restarts it and route and backend are then removed.

### Error log

from pod's logs:
```
2022-03-28T20:12:42+08:00    warn    ingress/ingress.go:231    sync ingress failed, will retry    {"object": {"Type":3,"Object":{"Key":"keycloak/keycloak","GroupVersion":"networking/v1","OldObject":null},"Tombstone":{}}, "error": "ingr │
│ ess.networking.k8s.io \"keycloak\" not found"}                                                                                                                                                                                              │
│ panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                     │
│ [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x18cb92c]                                                                                                                                                                    │
│                                                                                                                                                                                                                                             │
│ goroutine 174 [running]:                                                                                                                                                                                                                    │
│ github.com/apache/apisix-ingress-controller/pkg/ingress.(*ingressController).handleSyncErr(0xc0006b4a80, 0x198c020, 0xc00060e960, 0x1f9e100, 0xc000356fa0)                                                                                  │
│     /build/controller/pkg/ingress/ingress.go:237 +0x7ac                                                                                                                                                                                     │
│ github.com/apache/apisix-ingress-controller/pkg/ingress.(*ingressController).runWorker(0xc0006b4a80, 0x1fd9bf0, 0xc000358500)                                                                                                               │
│     /build/controller/pkg/ingress/ingress.go:81 +0xcc                                                                                                                                                                                       │
│ created by github.com/apache/apisix-ingress-controller/pkg/ingress.(*ingressController).run                                                                                                                                                 │
│     /build/controller/pkg/ingress/ingress.go:68 +0x230 ```
```
### Expected result

pod not to crash

**Comment by tao12345666333** [comment]
Thanks for your report. It has been fixed https://github.com/apache/apisix-ingress-controller/pull/872

It will be released in v1.5 this month.

You can use the latest code in the master branch for testing. Or wait for v1.5 to be released.

**Comment by hyt05** [comment]
Is there any way to fix it temporarily, our environment is now paralyzed, and the upgrade to 1.4.1 still reports this error

**Comment by AlinsRan** [comment]
1. yaml and version info
* apisix-ingress-controller version
```shell
$ apisix-ingress-controller version
apisix-ingress-controller version 1.4.1-no-git-module-go1.16.15
```
* ingress.yaml
```yaml
# ingress.yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: httpserver-ingress
spec:
  # apisix-ingress-controller is only interested in Ingress
  # resources with the matched ingressClass name, in our case,
  # it's apisix.
  ingressClassName: apisix
  rules:
  - host: httpbin.org
    http:
      paths:
      - backend:
          service:
            name: httpbin
            port:
              number: 80
        path: /
        pathType: Prefix
```

```shell
# create ingress resource
$ kubectl apply -f ingress.yaml

# verify
$ curl http://172.18.0.2:31091/ip -H 'Host:httpbin.org'
{
  "origin": "10.244.0.1"
}
# delete ingress resource
$ kukubectl delete ingress httpserver-ingress
ingress.networking.k8s.io "httpserver-ingress" deleted
# result
$ kubectl get pods -n $NS
NAME                                         READY   STATUS    RESTARTS   AGE
apisix-bfd7b7fc-bxljf                        1/1     Running   0          6m4s
apisix-etcd-0                                1/1     Running   0          6m4s
apisix-etcd-1                                1/1     Running   0          6m4s
apisix-etcd-2                                1/1     Running   0          6m4s
apisix-ingress-controller-6c4f7b6749-cmwvg   1/1     Running   0          6m4s
```

**Comment by AlinsRan** [comment]
> 1. yaml and version info
> 
> * apisix-ingress-controller version
> 
> ```shell
> $ apisix-ingress-controller version
> apisix-ingress-controller version 1.4.1-no-git-module-go1.16.15
> ```
> 
> * ingress.yaml
> 
> ```yaml
> # ingress.yaml
> apiVersion: networking.k8s.io/v1
> kind: Ingress
> metadata:
>   name: httpserver-ingress
> spec:
>   # apisix-ingress-controller is only interested in Ingress
>   # resources with the matched ingressClass name, in our case,
>   # it's apisix.
>   ingressClassName: apisix
>   rules:
>   - host: httpbin.org
>     http:
>       paths:
>       - backend:
>           service:
>             name: httpbin
>             port:
>               number: 80
>         path: /
>         pathType: Prefix
> ```
> 
> 2. Recurrence steps
> 
> ```shell
> # create ingress resource
> $ kubectl apply -f ingress.yaml
> 
> # verify
> $ curl http://172.18.0.2:31091/ip -H 'Host:httpbin.org'
> {
>   "origin": "10.244.0.1"
> }
> # delete ingress resource
> $ kukubectl delete ingress httpserver-ingress
> ingress.networking.k8s.io "httpserver-ingress" deleted
> # result
> $ kubectl get pods -n $NS
> NAME                                         READY   STATUS    RESTARTS   AGE
> apisix-bfd7b7fc-bxljf                        1/1     Running   0          6m4s
> apisix-etcd-0                                1/1     Running   0          6m4s
> apisix-etcd-1                                1/1     Running   0          6m4s
> apisix-etcd-2                                1/1     Running   0          6m4s
> apisix-ingress-controller-6c4f7b6749-cmwvg   1/1     Running   0          6m4s
> ```

Cannot be recurrence.

**Comment by tao12345666333** [comment]
@hyt05 hi, as @AlinsRan said, this issue has been resolved in v1.4.1.

If you do encounter this issue, please create a new issue with detailed steps to reproduce and results. thanks
